python/cuml/manifold/umap.pyx

```diff
@@ -419,8 +298,11 @@ cdef class UMAP:
                                             order = "C", dtype=np.float32))
         cdef uintptr_t embed_ptr = embedding.device_ctypes_pointer.value
 
+        cdef cumlHandle * handle_ = < cumlHandle * > < size_t > self.handle.getHandle()
+
         self.umap.transform(
-                        <float*>x_ptr,
+                       handle_[0],
+                       <float*>x_ptr,
                        <int>X_m.shape[0],
                        <int>X_m.shape[1],
                        <float*> self.embeddings,
```

Comment:
Same as above, missing a `cuml.sync()` before assigning the result to the dataframe and deleting X_m

Reply:
Same question as the knn below- in order for this function to complete properly, it would need to block the `cudaFree()` calls until after the `optimize_layout` has complete, otherwise it would be freeing the memory while the kernel is still writing to it.

Per the developer guide, we should expect that the cuml handle is not used across threads. Would cython be doing something here to make run `self.umap.transform()` asynchronously?

Reply:
had totally forgotten the bit of the cuml handle not being used among threads, so that for sure makes it so you're right I believe